### PR TITLE
improve coloring in plots tab - better UI, and enable coloring by clinical data

### PR DESCRIPTION
**Pull request by adamabeshouse**
image

Also, legends can now be clicked on to highlight data
image

image

### CODE REVIEW

**Comment by tmazor** [comment]
Awesome @adamabeshouse ! This is a great addition!

I do have a few thoughts -- everything below is using this query: https://www.cbioportal.org/results?=&session_id=5a1eef06498eb8b3d560f502&show_samples=false

1) Looks like when the page first loads, it's defaulting to color by both Mutation & Copy Number? I think we used to start with just Mutation by default?
2) "Mutation Type" should have an * after, to link to the footnote about driver annotation settings
3) If I color by Diagnosis Age, the legend just shows two colored dots, I assume these are the min & max? Can we instead show a continuous color spectrum? I think that's a common way to show a label for something like this.
4) If I then select "Log Scale", the text "(log scale)" appears somewhat awkwardly in the middle of the legend:
image
I can also see that some samples are colored slightly differently, but it's hard to appreciate without seeing a labeled continuous color gradient in the legend.
5) When the legend is horizontal and below the chart, can you add some space between the values? In this example, "NO" is closer to the green circle that means "YES" than it is to it's own red circle (plot: Mut# (log scaled) vs Dx, colored by Family History of Primary Brain Tumor):
image
6) For OS Status, the legend text is actually overlapping (plot: Mut# (log scaled) vs Dx, colored by OS Status):
image
7) This tooltip is cut off by the plot boundary (plot: Mut# (log scaled) vs Dx, colored by Prior Cancer Diagnosis Occurence)
image
8) When I color by a gene, the legend includes the gene name. When I color by a clinical attribute, that name should also be part of the legend.

**Comment by adamabeshouse** [comment]
@tmazor regarding (1), yes that's because @jjgao requested to default color by CNA now as well.

**Comment by adamabeshouse** [comment]
@tmazor okay, I made all of the requested changes, please give it another lookover when you can

**Comment by tmazor** [comment]
looks good to me @adamabeshouse !

File: src/pages/staticPages/webAPI/WebAPIPage.tsx [hunk 1]
```diff
@@ -119,10 +119,10 @@ export default class WebAPIPage extends React.Component<{}, {}> {
                 </Helmet>
 
                 <h1>Web API</h1>
```
Comment (alisman) [inline src/pages/staticPages/webAPI/WebAPIPage.tsx:121]: rebase

File: src/shared/cache/ClinicalDataCache.ts [hunk 2]
```diff
@@ -64,6 +67,14 @@ export function clinicalAttributeIsLocallyComputed(attribute: {
     );
 }
 
+export type ClinicalDataCacheEntry = {
```
Comment (alisman) [inline src/shared/cache/ClinicalDataCache.ts:70]: maybe a comment explaining why we include things like color state in data cache

File: src/shared/cache/ClinicalDataCache.ts [hunk 3]
```diff
@@ -259,15 +270,81 @@ export default class ClinicalDataCache extends MobxPromiseCache<
...
                         studyToMutationMolecularProfilePromise.result!,
                         studyIdToStudyPromise.result!,
                         coverageInformationPromise.result!
-                    ),
+                    );
+                    let categoryToColor;
+                    let numericalValueToColor;
+                    let logScaleNumericalValueToColor;
+                    let numericalValueRange;
+                    if (q.datatype === 'STRING') {
```
Comment (alisman) [inline src/shared/cache/ClinicalDataCache.ts:286]: can we move this out into a function

File: src/shared/components/oncoprint/OncoprintUtils.ts [hunk 4]
```diff
@@ -899,6 +903,8 @@ export function makeClinicalTracksMobxPromise(
                     }
                 } else if (attribute.datatype === 'STRING') {
                     ret.datatype = 'string';
+                    (ret as any).category_to_color =
```
Comment (alisman) [inline src/shared/components/oncoprint/OncoprintUtils.ts:906]: how come any?

File: src/shared/components/plots/BoxScatterPlot.tsx [hunk 5]
```diff
@@ -272,46 +282,99 @@ export default class BoxScatterPlot<
...
+
     private get legend() {
         if (this.props.legendData && this.props.legendData.length) {
             let legendData = this.props.legendData;
             if (this.legendLocation === 'bottom') {
                 // if legend is at bottom then flatten labels
                 legendData = legendData.map(x => {
-                    let name = x.name;
-                    if (Array.isArray(x.name)) {
+                    let { name, ...rest } = x;
```
Comment (alisman) [inline src/shared/components/plots/BoxScatterPlot.tsx:323]: neat.  this assigns everything but x.name to an object called "rest"?

File: src/shared/components/plots/PlotUtils.ts [hunk 6]
```diff
@@ -1,18 +1,121 @@
...
     IThreshold1D,
-    IWaterfallPlotData,
-    IScatterPlotData,
-    IBoxScatterPlotPoint,
     IThreshold2D,
-    IValue1D,
-    IValue2D,
 } from 'pages/resultsView/plots/PlotsTabUtils';
-import { IBoxScatterPlotData } from './BoxScatterPlot';
+import Timer = NodeJS.Timer;
```
Comment (alisman) [inline src/shared/components/plots/PlotUtils.ts:8]: What's Timer for?

File: src/shared/components/plots/BoxScatterPlot.tsx (hunk 5 above)
Comment (adamabeshouse) [reply src/shared/components/plots/BoxScatterPlot.tsx:323]: yep, im not sure the technical name for this but it rules

File: src/shared/components/oncoprint/OncoprintUtils.ts (hunk 4 above)
Comment (adamabeshouse) [reply src/shared/components/oncoprint/OncoprintUtils.ts:906]: typescript doesnt handle this well, union type

**Pull request merged**